src/sync/generic_network_transport.hpp

```diff
@@ -0,0 +1,335 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+
+#ifndef REALM_GENERIC_NETWORK_TRANSPORT_HPP
+#define REALM_GENERIC_NETWORK_TRANSPORT_HPP
+
+#include <functional>
+#include <string>
+#include <memory>
+#include <map>
+#include <vector>
+
+#include <realm/util/to_string.hpp>
+
+namespace realm {
+namespace app {
+namespace error {
+
+#pragma mark Errors
+
+struct AppError : public std::runtime_error
+{
+    enum class Type {
+        Generic,
+        Client,
+        Session,
+    };
+
+    AppError()
+    : std::runtime_error("AppError")
+    , type(Type::Generic)
+    {
+    };
+    AppError(std::string msg, Type classification = Type::Generic)
```

Comment:
This interface is breaking how errors are currently being reported. Currently, we are reporting this like this:  https://github.com/realm/realm-object-store/blob/master/src/sync/async_open_task.hpp#L45 using an exception or using the error code directly: https://github.com/realm/realm-object-store/blob/master/src/sync/sync_session.hpp#L140

SDK's can then use the error_category and error_code to map to SDK specific errors, e.g like Java is doing here: https://github.com/realm/realm-java/blob/master/realm/realm-library/src/main/cpp/io_realm_SyncSession.cpp#L163

I'm not sure exactly what the right API is here, but the App should IMO not to trying to remap errors. Instead, they should be delivered more or less directly as they are to SDK's which will then map them to more proper error types?

Reply:
But I suspect creating an error_category instead of a `Type` and then assign error_codes to all the app specific errors would fit the current pattern better?